frontend/src/metabase/dashboard/components/Dashboard/Dashboard.unit.spec.tsx

```diff
@@ -0,0 +1,188 @@
+        setSharing={jest.fn()}
+        setParameterValue={jest.fn()}
+        setEditingParameter={jest.fn()}
+        setParameterIndex={jest.fn()}
+        onUpdateDashCardVisualizationSettings={jest.fn()}
+        onUpdateDashCardColumnSettings={jest.fn()}
+        onReplaceAllDashCardVisualizationSettings={jest.fn()}
+        onChangeLocation={jest.fn()}
+        onSharingClick={jest.fn()}
+        onEmbeddingClick={jest.fn()}
+        toggleSidebar={jest.fn()}
+        closeSidebar={jest.fn()}
+        closeNavbar={jest.fn()}
+      />
+    );
+  }
+
+  const { rerender } = renderWithProviders(
+    <TestComponent
+      dashboard={opts.dashboard}
+      dashboardId={opts.dashboardId}
+      selectedTabId={opts.selectedTabId}
+      parameters={opts.parameters}
+      isEditing={opts.isEditing}
+    />,
+  );
+
+  if (!opts.skipLoader) {
+    await waitForLoaderToBeRemoved();
+  }
+
+  return {
+    rerender: (overrides: Partial<SetupOpts> = {}) =>
+      rerender(<TestComponent {...{ ...opts, ...overrides }} />),
+    mockLoadDashboardParams,
+    mockFetchDashboard,
+    mockFetchDashboardCardData,
+    mockFetchDashboardCardMetadata,
+  };
+}
+
+describe("Dashboard data fetching", () => {
+  afterEach(() => jest.clearAllMocks());
+
+  it("should fetch dashboard on first load", async () => {
+    const mocks = await setup();
+    expect(mocks.mockFetchDashboard).toHaveBeenCalledTimes(1);
+  });
+
+  it("should not fetch anything on re-render", async () => {
+    const mocks = await setup();
+    jest.clearAllMocks();
+    mocks.rerender();
+    expect(mocks.mockFetchDashboard).toHaveBeenCalledTimes(0);
+    expect(mocks.mockFetchDashboardCardData).toHaveBeenCalledTimes(0);
+    expect(mocks.mockFetchDashboardCardMetadata).toHaveBeenCalledTimes(0);
+  });
+
+  it("should fetch dashboard on dashboard id change", async () => {
+    const mocks = await setup();
+    mocks.rerender({ dashboardId: 20 });
+    expect(mocks.mockFetchDashboard).toHaveBeenCalledTimes(2);
+  });
+
+  it("should fetch card data and metadata when tab id changes", async () => {
+    const mocks = await setup();
+    mocks.rerender({ selectedTabId: 1 });
+    expect(mocks.mockFetchDashboardCardData).toHaveBeenCalledTimes(1);
+    expect(mocks.mockFetchDashboardCardMetadata).toHaveBeenCalledTimes(1);
+  });
+
+  it("should fetch card data when dashboard changes to non-empty", async () => {
+    const mocks = await setup({
+      dashboardId: null,
+      dashboard: null,
+      skipLoader: true,
+    });
+    expect(mocks.mockFetchDashboardCardData).toHaveBeenCalledTimes(0);
+
+    mocks.rerender({
+      dashboardId: null,
+      dashboard: createMockDashboard({ id: 20 }),
+    });
+    expect(mocks.mockFetchDashboardCardData).toHaveBeenCalledTimes(1);
+  });
+
+  it("should fetch card data when isDirty changes and not editing", async () => {
+    const mocks = await setup({ dashboard: dashboardWithCards() });
+    jest.clearAllMocks();
```

Comment:
A few tests are using `setup()` + `jest.clearAllMocks()` pattern. Would it make sense to run `jest.clearAllMocks()` at the end of `setup`?

Reply:
I don't think so. IMO tests must not be too abstract, and this is quite an important detail of each test case to be hidden somewhere.